Removed scope pooling and added diagnostics

davidfowl opened the pull request:
- The ConcurrentQueue showed a 48% regression on AMD scenarios and a 4% regression on citrine
linux.
- Added event source event to determine how many scopes services and disposable services are resolved

Fixes https://github.com/dotnet/aspnetcore/issues/31703

Citrine:

```
1,950,852  -> 2,020,030 
```

AMD

```
1,975,517-> 3,986,695 
```

davidfowl committed:
Use a thread static instead of concurrent queue for the scope pool
- The ConcurrentQueue showed a 48% regression on AMD scenarios and a 4% regression on citrine
linux.
- Removed interface dispatch on clear

jkotas commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs:
```diff
@@ -5,40 +5,42 @@
-        private readonly ConcurrentQueue<State> _queue = new();
+        // Stash the scope state on a thread static
+        [ThreadStatic]
+        private static State t_tlsScopeState;
```
What is the maximum size of this State object that will be stored here?

davidfowl replied:
I'd expect an extra large dictionary to be ~400 empty entries (that's what we saw in the orchard dumps).

jkotas replied:
Is there an upper limit? The TLS caches like this typically have to have upper bond to prevent unbounded memory leak. Or they have to be kept in a weak handle so that they can get collected.

davidfowl replied:
I assume you're worried about the size of the entries empty entries on thread pool threads never being cleaned up?

jkotas replied:
Right. 

For example, the dreadful StringBuilderCache has upper limit of 360 for the same reason: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Text/StringBuilderCache.cs#L12

davidfowl replied:
> Is there an upper limit in the worse case? 

I can add one, it's related to the comment here https://github.com/dotnet/runtime/blob/7fb611ec5aa9ef9c7080b20a3a386481f873bff6/src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs#L65. I'm not sure it matters though because in the most common cases, you end up constantly (per request) allocating a dictionary with that many entries anyways. It'll never get bigger than what your code was trying to use. I fear if we throw away entries that are too big then they'll be repeatedly allocated and thrown out. I could add a worst cast just in case (some extremely high number just to avoid pooling it at all).

> The TLS caches like this typically have to have upper bound to prevent unbounded memory leak. 

Aren't you intrinsically throttled by the number of threads? In practice, I don't see how this could lead to an unbounded memory leak.

jkotas replied:
What about heteregonous requests? If you have one rarely executed request type that needs a lot of entries, all threads are eventually going to get the large collection attached to them and most of it will be useless.

>  I fear if we throw away entries that are too big then they'll be repeatedly allocated and thrown out. 

I do not see a problem with it. This optimization will not kick in and I think it is perfectly fine.

davidfowl replied:
> What about heterogonous requests? If you have one rarely executed request type that needs a lot of entries, all threads are eventually going to get the large collection attached to them and most of it will be useless.

Definitely possible, scopes can be used in the same app for background tasks and those might look different per request. I don't have any data but experience tells me the number of scoped services per request of the typical app is very similar. That's why I like the trimming capacity solution much better than throwing out the entire state object. It'll at the very least save initial resizes on other requests.

davidfowl replied:
I added a size. Not sure if it's too big and it's not a prime or power of 2 so it feels wrong 😄

jkotas replied:
This makes it up to 24kB cached per thread if I have done the math right. Is it correct?

davidfowl replied:
How did you get 24KB? I've been trying to work out the size by looking at the size of an entry assuming it's also disposable:

![image](https://user-images.githubusercontent.com/95136/114520508-5e393680-9bf6-11eb-87d4-3cdc37811171.png)

jkotas replied:
Actually, I think it is 36kB:
```
4 hashCode
4 int next

16 ServiceCacheKey (Type, Slot, padding)

8 object value

4 bucket
```

davidfowl replied:
Too big? 😄 I know people set large thread pool limits (1000s) but I think 1000 scoped services per request is pretty extreme, you probably have lots of memory anyways 😄

jkotas replied:
Yes, it feels too big to me.

davidfowl replied:
OK I made some changes in the latest PR

davidfowl committed:
Add a maximum size

davidfowl committed:
Added comment about the size and removed null tlsCache check on return

watfordgnf commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs:
```diff
@@ -5,40 +5,45 @@
 
-        private int _count;
-        private readonly ConcurrentQueue<State> _queue = new();
+        // More than 1000 scopes services seems like too many to cache (~36KB in size)
```
```suggestion
        // More than 1000 scoped services seems like too many to cache (~36KB in size)
```

davidfowl committed:
Made some adjustments based on PR feedback
- Cache 100 scoped entries maximum
- Keep track of the maximum capacity up to a limit

gfoidl commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs:
```diff
@@ -43,6 +49,9 @@ public bool Return(State state)
                 return true;
             }
 
+            // Clamp the size between the min and max
+            _maxScopeSize = Math.Min(MaxCapacity, Math.Max(size, _maxScopeSize));
```
```suggestion
            _maxScopeSize = Math.Clamp(size, _maxScopeSize, MaxCapacity);
```

davidfowl replied:
Didn't it because it's not on .NET Standard 2.0 (and I didn't want to cross compile).

gfoidl commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs:
```diff
@@ -14,27 +14,33 @@ internal class ScopePool
         public bool Return(State state)
         {
-            if (state.Size <= MaxResolvedServicesCount)
+            var size = state.Size;
```
Here 😉 
```suggestion
            int size = state.Size;
```

davidfowl commented:
Unfortunately this regresses https://github.com/dotnet/runtime/issues/47607 

![image](https://user-images.githubusercontent.com/95136/114601567-7dfa4a00-9c4a-11eb-9929-68f9450163f2.png)

jkotas commented:
As we have discussed offline, the pooling here has problems that include regressing performance of heteregenous services that process requests with varying number of services, or increasing GC pause times due to introducing Gen2 -> Gen0 references.

I am sorry I cannot signoff on this change. I believe that the use of pooling here is a bad idea and it should be reverted.

Feel free to overrule me and get somebody else to sing-off on it.

davidfowl commented:
Thanks for the feedback @jkotas, my take is that infrastructure shouldn't cost more than the work it's supporting. The change was motivated by the fact that orchard pays for the overhead of DI as being the top allocation.

I think I've struck a reasonable balance based on your feedback and have adjusted numbers appropriately. I think one last tweak would be to tweak the maximum capability so the worst case behavior of completely heterogeneous scopes is reasonable (but I think that scenario is rare). In a typical ASP.NET Core app, the scopes are per request and are similar in size. The trick here is to find limit that cover reasonable size apps but don't harm other use cases. I think ~50 scoped services and a max of 100 initialCapacity is good for this (I will change the current numbers).

JamesNK commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopePool.cs:
```diff
@@ -5,40 +5,54 @@
-                return false;
+                // Stash the state back in TLS
+                state.Clear();
+                t_tlsScopeState = state;
```
Rent and Return won't always run on the same thread, right? You could have two scopes in progress and after async/await both return their scopes on the same thread, with the second returned scope overwriting the first.

davidfowl replied:
Yep, but I don't see why that matters.

jkotas replied:
It just means that the hit rate of the pool will vary based on how perfectly distributed the requests are. If you are lucky, close to everything will hit the pool. If you are very unlucky, close to nothing will hit the pool.

davidfowl replied:
I'm a glass always full type of person.

davidfowl commented:
OK I did a bunch of work here to figure out if the GC is a problem in the original issue where this was filed. Initially there was an allocation profile only but without a matching CPU profile or GC trace to see whether it made sense to remove allocations. A CPU profile on the same benchmark shows this:

![image](https://user-images.githubusercontent.com/95136/114683049-898a5700-9cc4-11eb-8a9a-bcee359fc411.png)

There's 1.8% spent in the gc1 function (I have no idea what this does specifically cc @Maoni0).

![image](https://user-images.githubusercontent.com/95136/114683444-e2f28600-9cc4-11eb-8995-953fa57230d4.png)

1.8% isn't the end of the world in the grand scheme of things but orchard allocates a crapload of things per request and we started to look at reducing this *and* DI is one of the top allocations (see https://github.com/dotnet/runtime/issues/47607). It's possible that orchard is "mis-uing" DI but this felt like something we should try to tackle. If that helps with the motivation.

jkotas commented:
> There's 1.8% spent in the gc1 function

Does this pooling change actually decrease time spend in the GC? The Gen2 -> Gen0 references that this pooling introduces adds more work that the GC has to do. It is not obvious to me that the reduced allocation will pay for this extra work.

davidfowl commented:
> Does this pooling change actually decrease time spent in the GC? The Gen2 -> Gen0 references that this pooling introduces add more work that the GC has to do. It is not obvious to me that the reduced allocation will pay for this extra work.

Looks worse, the number of requests isn't exactly the same, but the time looks split now between the plan phase and gc1

![image](https://user-images.githubusercontent.com/95136/114744956-118f5180-9d03-11eb-85a6-20d9c69a6a15.png)

davidfowl committed:
Remove pooling, just track sizes

davidfowl commented:
I removed the pooling and am currently testing out this change that keeps track of the maximum number of services resolved per scope (up to some limit).

Maoni0 commented:
@davidfowl the profiles you showed seem odd... gc1 is the method that actually does the GC internal work and it mostly just calls other methods (that aren't inlined) so you'd see exc is usually 0 and inc is some non zero number. but yours just show the same for both... you might want to expand it callees and see what's up.

if you want to see how much work GC is doing due to the generational aspect, ie, marking objects because they are held live by objects in the generations that aren't getting collecting, [this](https://github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md#figuring-out-the-amount-of-work-for-ephemeral-gcs) explains how to do that.

davidfowl committed:
Remove pre-sizing logic
- Added an event that we can ask customers to turn on to see if scope in real applications are mostly homogeneous

davidfowl committed:
Add the hashcode of the tracker so that we can detect scope sizes per container instance.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs:
```diff
@@ -14,22 +14,20 @@ internal sealed class ServiceProviderEngineScope : IServiceScope, IServiceProvid
-        // the list of disposable entries only, since ResolvedServices is a concurrent dictionary.
-        // For other scopes, it protects ResolvedServices and the list of disposables
-        private readonly object _scopeLock = new object();
+        private ScopeTracker.State _state;
```
`readonly` ?

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/src/ScopeTracker.cs:
```diff
@@ -0,0 +1,46 @@
+            public List<object> Disposables { get; set; }
+
+            public int DisposableServicesCount => Disposables?.Count ?? 0;
+            public int ResolvedServicesCount => ((Dictionary<ServiceCacheKey, object>)ResolvedServices).Count;
```
Is this cast safe? Or do we only expect `ResolvedServicesCount` to be called on non-root scopes?

davidfowl replied:
IT used to be safe, but now it's not.

davidfowl replied:
Removed cast

davidfowl committed:
Simplify

davidfowl committed:
Remove cast

eerhardt reviewed (approved):
LGTM - I like that we have the ability to get more information from the system now. It should help in diagnosis in the future.

davidfowl commented:
> LGTM - I like that we have the ability to get more information from the system now. It should help in diagnosis in the future.

Ran it on orchard:


```
HasStack="True" ThreadID="18,635" ProcessorNumber="0" serviceProviderHashCode="32,961,781" scopedServicesResolved="209" disposableServices="7"
HasStack="True" ThreadID="18,619" ProcessorNumber="0" serviceProviderHashCode="32,961,781" scopedServicesResolved="209" disposableServices="7"
HasStack="True" ThreadID="18,635" ProcessorNumber="0" serviceProviderHashCode="53,813,228" scopedServicesResolved="0" disposableServices="0"
```

davidfowl commented:
Thanks for all the feedback folks! I learned alot

davidfowl commented:
@Maoni0 I still have the traces so I'll spend some time looking at them and let you know what I find.

gfoidl commented:
> let you know what I find.

Please do this here (I'd like to learn too...)
